### PR TITLE
[WIP] e2e: add retry mechansim to pick nodeport

### LINKED ISSUE #125466: [Flaking test] [sig-network] Services should release NodePorts on delete
### Which jobs are flaking?

pull-kubernetes-e2e-kind-ipv6

### Which tests are flaking?

[sig-network] Services should release NodePorts on delete [link](https://github.com/kubernetes/kubernetes/blob/1815a14c32a3f87e1146b21a04eef796ccc079cf/test/e2e/network/service.go#L1664)

### Since when has it been flaking?

Looks line 7th June, but maybe we were just lucky before:
image

### Testgrid link

https://testgrid.k8s.io/presubmits-kubernetes-blocking#pull-kubernetes-e2e-kind-ipv6

### Reason for failure (if possible)

The logs are like:
```
{ failed [FAILED] failed to create service: nodeport-reuse in namespace: services-2468: Service "nodeport-reuse" is invalid: spec.ports[0].nodePort: Invalid value: 30558: provided port is already allocated
In [It] at: k8s.io/kubernetes/test/e2e/network/service.go:1724 @ 06/12/24 09:19:41.932
```

### Anything else we need to know?

_No response_

### Relevant SIG(s)

/sig network

### PR DESCRIPTION
**Pull request by aojea**
/kind flake

Following up on https://github.com/kubernetes/kubernetes/pull/125597 because test freeze will be soon and we'll not have time to merge after it.

Keeping @varshith257 as author and greatly thanking him for working on this

```release-note
NONE
```

Fixes: #125466, #126400

### CODE REVIEW

**Review by tnqn — COMMENTED** [review]
/lgtm

**Comment by varshith257** [comment]
Thanks, @aojea for following up on this

**Comment by danwinship** [comment]
> Fixes: #125466, #126400

no, #126400 is failing in a different test

**Comment by varshith257** [comment]
> no, #126400 is failing in a different test

```
[FAILED] failed to update Service "external-local-update": Internal error occurred: failed to allocate requested HealthCheck NodePort 30202: provided port is already allocated
In [It] at: k8s.io/kubernetes/test/e2e/network/loadbalancer.go:1616 @ 07/21/24 12:41:29.957
```

From logs we can clearly see that failure is due to nodePort allocation. From this PR we achieving to add retry mechanism to allocate other port if existing port is used by other services.


@danwinship Do you any suggestions or ideas on mind ?

File: test/e2e/network/service.go [hunk 1]
```diff
@@ -1684,9 +1684,24 @@ var _ = common.SIGDescribe("Services", func() {
 		service.Spec.Type = v1.ServiceTypeNodePort
 
 		ginkgo.By("creating service " + serviceName + " with type NodePort in namespace " + ns)
-		service, err := t.CreateService(service)
-		framework.ExpectNoError(err, "failed to create service: %s in namespace: %s", serviceName, ns)
-
+		// Retry mechanism for picking a node port between 30000 and 30085
+		// to avoid conflicts
+		// https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/3668-reserved-service-nodeport-range
```
Comment (danwinship) [inline test/e2e/network/service.go:1689]: Use `e2eservice.NodePortRange`. (It's still hardcoded, but at least that way it's only hardcoded in one place.)

Also, the comment doesn't really explain what you're doing very well (and I initially thought you were fixing the bug in the wrong place).

"For the test to work, we need to make sure the Service's NodePort doesn't get reallocated to another Service after service is deleted. So we have to find an unused NodePort in the static allocation range."

File: test/e2e/network/service.go [hunk 2]
```diff
@@ -1684,9 +1684,24 @@ var _ = common.SIGDescribe("Services", func() {
...
+		const maxRetries = 5
+		created := false
+		for i := 0; i < maxRetries; i++ {
+			service.Spec.Ports[0].NodePort = int32(30000 + rand.Intn(86))
+			createdService, err := t.CreateService(service)
+			if err == nil {
```
Comment (danwinship) [inline test/e2e/network/service.go:1695]: maybe only retry on `apierrors.IsInvalid()`?
(maybe `retry.OnError(retry.DefaultRetry, apierrors.IsInvalid, func() error { ... })`?)

**Review by danwinship — CHANGES_REQUESTED** [review]

**Comment by danwinship** [comment]
> From logs we can clearly see that failure is due to nodePort allocation.

(HealthCheckNodePort actually; not _exactly_ the same thing.)

> From this PR we achieving to add retry mechanism to allocate other port if existing port is used by other services.

This PR changes the _initial_ NodePort allocation to use a port in the static range, so as to guarantee that when we delete the service later, it's not possible that the NodePort is used by other services.

Though of course, once we fix #126400, that guarantee would be broken; if this test and the LB HCNP test ran in parallel-ish, then you could get:

- `"should release NodePorts on delete"`: create service with randomly-selected NodePort 32123 → succeeds because nobody is using that port
- `"should release NodePorts on delete"`: delete service
- `"should handle updates to ExternalTrafficPolicy field"`: create service with randomly-selected HealthCheckNodePort 32123 → succeeds because nobody is using that port
- `"should release NodePorts on delete"`: recreate service with NodePort 32123 → **FAIL**
- `"should handle updates to ExternalTrafficPolicy field"`: delete service
- `"should handle updates to ExternalTrafficPolicy field"`: recreate service with HealthCheckNodePort 32123 → succeeds because nobody is using that port

File: test/e2e/network/service.go (hunk 2 above)
Comment (aojea) [reply test/e2e/network/service.go:1695]:
```
Response Status: 422 Unprocessable Entity in 7 milliseconds
The Service "nodeport-service2" is invalid: spec.ports[0].nodePort: Invalid value: 30101: provided port is already allocated
```
👍

File: test/e2e/network/service.go (hunk 1 above)
Comment (aojea) [reply test/e2e/network/service.go:1689]: 👍

**Comment by aojea** [comment]
hmm, the reported issue was also on the loadbalancer tests, and these are the service tests 

this is more involved than I thought

**Comment by danwinship** [comment]
> hmm, the reported issue was also on the loadbalancer tests, and these are the service tests

#125466 is on the service tests (and this PR would fix it). #126400 is on the LB tests, and a similar fix could be made there (and, in isolation, would fix that test), but if you fix both tests in the same way then it gets un-fixed as described above, because both tests might end up picking the same "reserved" nodeport to use.

I think the fix is either:
1. Allocate specific nodeports to different tests statically. So, eg, the service test would always use `e2eservice.NodePortRange.Base + 12` and the LB test would always use `e2eservice.NodePortRange.Base + 37`. We sort of do this with hostports / hostNetwork pod ports, though it's very ad hoc. But this seems less likely to work well with nodeports because there are fewer of them so we're more likely to end up conflicting with something that's an existing part of the cluster the e2e tests are running in...
2. Keep track of reserved nodeports at runtime, `e2eservice.ReserveNodePort()` or something (so then there'd also have to be an `e2eservice` function to request an unused nodeport too). This would be simpler if it was always literally `.NodePort`, but sometimes it's `.HealthCheckNodePort`.

probably 2?

**Comment by aojea** [comment]
definitively 1. is not possible because of the reasons you mention 

I think we should iterate in 2.

**Comment by aojea** [comment]
/close

There seems to be some volunteers https://github.com/kubernetes/kubernetes/issues/126400#issuecomment-2301661952

**Pull request closed without merging**